test/integration/connection-monitoring-and-pooling/connection_pool.test.ts

```diff
@@ -64,5 +65,82 @@ describe('Connection Pool', function () {
         });
       });
     });
+
+    describe(
+      'ConnectionCheckedInEvent',
+      { requires: { mongodb: '>=4.4', topology: 'single' } },
+      function () {
+        let client: MongoClient;
+
+        beforeEach(async function () {
+          await configureFailPoint(this.configuration, {
+            configureFailPoint: 'failCommand',
+            mode: 'alwaysOn',
+            data: {
+              failCommands: ['insert'],
+              blockConnection: true,
+              blockTimeMS: 500
+            }
+          });
+
+          client = this.configuration.newClient();
+          await client.connect();
+          await Promise.all(Array.from({ length: 100 }, () => client.db().command({ ping: 1 })));
+        });
+
+        afterEach(async function () {
+          await clearFailPoint(this.configuration);
+          await client.close();
+        });
+
+        describe('when a MongoClient is closed', function () {
+          it(
+            'a connection pool emits checked in events for closed connections',
+            { requires: { mongodb: '>=4.4', topology: 'single' } },
+            async () => {
+              const allClientEvents = [];
+              const pushToClientEvents = e => allClientEvents.push(e);
+
+              client
+                .on('connectionCheckedOut', pushToClientEvents)
+                .on('connectionCheckedIn', pushToClientEvents)
+                .on('connectionClosed', pushToClientEvents);
+
+              const inserts = Promise.allSettled([
+                client.db('test').collection('test').insertOne({ a: 1 }),
+                client.db('test').collection('test').insertOne({ a: 1 }),
+                client.db('test').collection('test').insertOne({ a: 1 })
+              ]);
+
+              // wait until all pings are pending on the server
+              while (allClientEvents.filter(e => e.name === 'connectionCheckedOut').length < 3) {
+                await sleep(1);
+              }
+
+              const insertConnectionIds = allClientEvents
+                .filter(e => e.name === 'connectionCheckedOut')
+                .map(({ address, connectionId }) => `${address} + ${connectionId}`);
+
+              await client.close();
+
+              const insertCheckInAndCloses = allClientEvents
+                .filter(e => e.name === 'connectionCheckedIn' || e.name === 'connectionClosed')
+                .filter(({ address, connectionId }) =>
+                  insertConnectionIds.includes(`${address} + ${connectionId}`)
+                );
+
+              expect(insertCheckInAndCloses).to.have.lengthOf(6);
+
+              // check that each check-in is followed by a close (not proceeded by one)
+              expect(insertCheckInAndCloses.map(e => e.name)).to.deep.equal(
+                Array.from({ length: 3 }, () => ['connectionCheckedIn', 'connectionClosed']).flat(1)
+              );
```

Comment:
suggest: can we just deep assert on the events, like the UTR does?

```
expect(
  allClientEvents.map(event => event.name)
).to.deep.equal(
  [checkout, checkout, checkout, checkIn, close, checkIn, close, checkIn, close]
)
```

Or is there something that asserting on the actual connection ids captures here that my proposed assertion doesn't?

Reply:
There's no assertions on the ids they're just filtering for the insert related checkins/outs/closes. There's more events because close is running endSessions and there's a connection created for the initial ping. So there's just a handful of things that aren't relevant to the test and would need revising if there were unrelated details changed about the other operations run